ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsReader.scala

```diff
@@ -0,0 +1,132 @@
+)(implicit ec: ExecutionContext)
+    extends LedgerDaoEventsReader {
+
+  protected val dbMetrics: metrics.daml.index.db.type = metrics.daml.index.db
+
+  override def getEventsByContractId(contractId: ContractId, requestingParties: Set[Party])(implicit
+      loggingContext: LoggingContext
+  ): Future[GetEventsByContractIdResponse] = {
+
+    val dbMetric: DatabaseMetrics = dbMetrics.getEventsByContractId
+
+    val eventProjectionProperties = EventProjectionProperties(
+      verbose = true, // Used by LfEngineToApi
+      witnessTemplateIdFilter = requestingParties
+        .map(_ -> Set.empty[Identifier])
+        .toMap, // Needed to get create arguments mapped
+      witnessInterfaceViewFilter = Map.empty, // We do not need interfaces mapped
+    )
+
+    for {
+      rawEvents <- dbDispatcher.executeSql(dbMetric)(
+        eventStorageBackend.eventReaderQueries.fetchContractIdEvents(
+          contractId,
+          requestingParties = requestingParties,
+        )
+      )
+
+      filteredRawEvents = rawEvents.filter(
+        _.event.witnesses.exists(requestingParties.map(identity))
+      )
+
+      deserialized <- Future.traverse(filteredRawEvents) {
+        _.event.applyDeserialization(lfValueTranslation, eventProjectionProperties)
+      }
+
+    } yield {
+      GetEventsByContractIdResponse(deserialized)
+    }
+  }
+
+  override def getEventsByContractKey(
+      contractKey: Value,
+      templateId: Ref.Identifier,
+      requestingParties: Set[Party],
+      maxEvents: Int,
+      startExclusive: Offset,
+      endInclusive: Offset,
+  )(implicit loggingContext: LoggingContext): Future[GetEventsByContractKeyResponse] = {
+    val dbMetric: DatabaseMetrics = dbMetrics.getEventsByContractKey
+    val keyHash: String = platform.Key.assertBuild(templateId, contractKey).hash.bytes.toHexString
+
+    val eventProjectionProperties = EventProjectionProperties(
+      // Used by LfEngineToApi
+      verbose = true,
+      // Needed to get create arguments mapped
+      witnessTemplateIdFilter = requestingParties.map(_ -> Set.empty[Identifier]).toMap,
+      // We do not need interfaces mapped
+      witnessInterfaceViewFilter = Map.empty,
+    )
+
+    for {
+
+      (rawEvents, prunedOffset) <- dbDispatcher.executeSql(dbMetric) { conn =>
+        def lookupSeq(o: Offset) = eventStorageBackend
+          .maxEventSequentialIdOfAnObservableEvent(o)(conn)
+          .getOrElse(EventSequentialId.beforeBegin)
+
+        val startSeqExclusive = lookupSeq(startExclusive)
+        val endSeqInclusive = lookupSeq(endInclusive)
+
+        val prunedOffset = parameterStorageBackend.prunedUpToInclusive(conn)
+
+        val rawEvents = eventStorageBackend.eventReaderQueries.fetchContractKeyEvents(
+          keyHash,
+          requestingParties,
+          maxEvents,
+          startSeqExclusive,
+          endSeqInclusive,
+        )(conn)
+
+        (rawEvents, prunedOffset)
+      }
+
+      filteredRawEvents = rawEvents.filter(
```

Comment:
jvm filtration complicates the pagination functionality: could we drop maybe filtration? (admin endpoint etc)